Implement the antctl framework

weiqiangt opened the pull request:
Resolves #203.

jianjuns commented on Makefile:
```diff
@@ -1,8 +1,10 @@
 LDFLAGS         :=
 GOFLAGS         :=
 BINDIR          := $(CURDIR)/bin
+ANTCTL_BINDIR   := $(CURDIR)/antctl-bin
```
Should we just use bin dir (and for the container image put windows and MAC bins in different dirs)?
@antoninbas @tnqn : what are you opinions?

tnqn replied:
I think the bin dir is good, not common to have separate bin dirs.

weiqiangt replied:
OK, I would remove this dir. I think we don't need to care windows and macOS build since we would not build `antctl` in any image release.

jianjuns replied:
Actually I still like some way to publish the Windows and MAC versions, and thought packaging them into the image could be one way. We could also provide download from github on the same time.

jianjuns replied:
I want not to complicate your current PR further, so we could consider Windows and MAC builds later.

weiqiangt replied:
Created #312 to track this issue.

jianjuns commented on build/yamls/base/antctl.yml:
```diff
@@ -0,0 +1,32 @@
+  name: antctl
+rules:
+  - nonResourceURLs:
+      - /apis/antctl.antrea.io/v1
```
Could we use some generic naming instead of indicating this is for antctl only? Like clusterinfo, or controllerinfo/agentinfo or nodeinfo?
@tnqn: what you think?

tnqn replied:
Agree, they should be named with what they are, instead of who use them.

weiqiangt replied:
Thanks guys, but could you give me some suggestions for the name?
This endpoint was used for accessing antctl server side and we will add commands not related to info, like `support-bundle`.

jianjuns replied:
I think we could use clusterinfo or now (or controllerinfo and agentinfo if we want to separate controller and agent/node).

Not sure about support-bundle. What API groups K8s uses for logging/metrics, etc.?

weiqiangt replied:
Changed to `clusterinfo.antrea.tanzu.vmware`.
The `metrics` in Kubelet does not have API group, it's a bare endpoint.

jianjuns commented on cmd/antctl/main.go:
```diff
@@ -0,0 +1,57 @@
+var rootCmd = &cobra.Command{
+	Use:   commandName,
+	Short: commandName + " is the command line tool for Antrea",
+	Long:  commandName + " is the command line tool for Antrea that supports showing the current states of Controller or Agent.",
```
"showing state of" or "showing runtime information of".

weiqiangt replied:
Thanks, changed.

jianjuns commented on cmd/antrea-agent/agent.go:
```diff
@@ -127,7 +128,16 @@ func run(o *Options) error {
 
 	go agentMonitor.Run(stopCh)
 
+	serverOpt, err := antctl.NewServerOptions(antctl.Definition, agentMonitor, nil, antctl.DefaultAgentLocalPort)
```
Joe suggested to use USD for local endpoints for security consideration. But we could address that in a new PR later.

weiqiangt replied:
I also prefer to use a USD. Since the agent is a daemon set, the port would be exposed directly on host.

weiqiangt replied:
@jianjuns, I have changed to use UDS, PTAL.

jianjuns commented on pkg/antctl/commandbundle.go:
```diff
@@ -0,0 +1,140 @@
+}
+
+// ApplyToRouter applies the CommandBundle to a router.
+func (b *CommandBundle) ApplyToRouter(r *mux.PathRecorderMux, aq monitor.AgentQuerier, cq monitor.ControllerQuerier) {
```
Sorry but is "router" a generic term from somewhere? If not maybe add some comments to explain what is a router.

weiqiangt replied:
My bad, I think `mux` would be better.

jianjuns commented on pkg/antctl/commandoption.go:
```diff
@@ -0,0 +1,344 @@
+
+var groups = map[commandGroup]*cobra.Command{}
+
+// Command returns the string representation of the command group.
```
Command -> String?

weiqiangt replied:
Fixed.

jianjuns commented on pkg/antctl/commandoption.go:
```diff
@@ -0,0 +1,344 @@
+	"github.com/vmware-tanzu/antrea/pkg/antctl/handlers"
+)
+
+type commandGroup uint
```
What is a commandGroup? Add some comments here?

weiqiangt replied:
Added some comments for it.

jianjuns commented on pkg/antctl/handlers/version.go:
```diff
@@ -0,0 +1,53 @@
+type VersionResponse struct {
+	AgentVersion      string `json:"agentVersion,omitempty" yaml:"agentVersion,omitempty"`
+	ControllerVersion string `json:"controllerVersion,omitempty" yaml:"controllerVersion,omitempty"`
+	AntctlVersion     string `json:"antctlVersion,omitempty" yaml:"antctlVersion,omitempty"`
```
This should not replied from the server, and so should not be defined in VersionResponse?

weiqiangt replied:
In the `definition.go`, the `versionTransform` will populate the `antctl` version to it.

weiqiangt replied:
`omitempty` makes sure the serialization result would not contains keys with empty values.

jianjuns replied:
I saw you populate antctl version, but again where we define fields not implemented by the server to the response message? Here we might just include a single version in the response message definition (for controller it is the controller version and for agent it is the agent version)?

weiqiangt replied:
OK, got your point, I have updated the code and shrunk the VersionResponse.

jianjuns commented on pkg/antctl/handlers/version.go:
```diff
@@ -0,0 +1,53 @@
+
+// VersionResponse describes the response of the version command.
+type VersionResponse struct {
+	AgentVersion      string `json:"agentVersion,omitempty" yaml:"agentVersion,omitempty"`
```
If this is returned by the server, it should include a single version field? Not sure if I misunderstood.

weiqiangt replied:
Replied in prior comments, thanks.

mengdie-song commented on build/yamls/antrea.yml:
```diff
@@ -313,6 +318,22 @@ spec:
+    app: antrea
+  name: v1.antctl.antrea.io
+spec:
+  group: antctl.antrea.io
```
The group here should be antctl.antrea.tanzu.vmware.com

weiqiangt replied:
Changed.

mengdie-song commented on cmd/antctl/main.go:
```diff
@@ -0,0 +1,57 @@
+	antctl.Init(rootCmd)
+
+	if err := rootCmd.Execute(); err != nil {
+		os.Exit(1)
```
Also flush logs before os.Exit.

weiqiangt replied:
Great catch, thanks.

mengdie-song commented on cmd/antrea-agent/agent.go:
```diff
@@ -127,7 +128,16 @@ func run(o *Options) error {
 	<-stopCh
+
 	klog.Info("Stopping Antrea agent")
+	<-serverStoppedCh // check if the antctl server stopped
```
I think that shutting down LocalServer by adding a defer clause here can avoid introducing serverStoppedCh here.

weiqiangt replied:
`serverStoppedCh` here is basically a `Future` like stuff. Since the agent will definitely be killed by the second signal, I use the channel here to ensure the server does have a `graceful` stop in best efforts.

mengdie-song replied:
Any reason that the server cannot be the same as the others (cniServer, nodeRouteController, networkPolicyController and agentMonitor)? Can we initialize it first, then run it in a separate go routine?

weiqiangt replied:
OK, I removed the wait of `stoppedCh` and we may need to discuss how to uniform them in future.

mengdie-song commented on pkg/antctl/server.go:
```diff
@@ -0,0 +1,50 @@
+	router := mux.NewPathRecorderMux("antctl")
+	opts.CommandBundle.ApplyToRouter(router, opts.AgentQuerier, opts.ControllerQuerier)
+	server := &http.Server{Handler: router}
+	stoppedCh := make(chan struct{})
```
Maybe you can remove the stoppedCh here and if we detect error from server.Serve, we can close stopCh if you want to exit main in this case.

weiqiangt replied:
`close(stopCh)` just signals components to stop, but no one knowns when they stopped. The `stoppedCh` here can tell the invoker it exits.

jianjuns commented on Makefile:
```diff
@@ -75,6 +76,19 @@ docker-tidy: $(DOCKER_CACHE)
 	GOBIN=$(BINDIR) $(GO) install $(GOFLAGS) -ldflags '$(LDFLAGS)' github.com/vmware-tanzu/antrea/cmd/...
 
+# TODO: strip binary when building realses
+ANTCTL_BINARIES := antctl-darwin antctl-linux antctl-windows
```
Even we do not want to include Windows and MAC bins now, we should include the Linux version to the container image. Does it mean we should build antctl with "make bin", and also have a way to build it with a Dockerfile (like other bins)?

weiqiangt replied:
`github.com/vmware-tanzu/antrea/cmd/...` in `make bin` covers all executable targets include `antctl`.

weiqiangt commented on pkg/antctl/commandbundle.go:
```diff
@@ -0,0 +1,140 @@
+
+func (b *CommandBundle) applyToRootCommand(root *cobra.Command, isAgent bool) {
+	defaultPath := filepath.Join(homedir.HomeDir(), ".kube", "config")
+	root.PersistentFlags().BoolP("debug", "d", false, "enable debug mode")
```
Make sense, changed to `-v` and `--verbose`.

antoninbas commented on pkg/antctl/client.go:
```diff
@@ -0,0 +1,126 @@
+	return u.RequestURI()
+}
+
+// Client does request to a CLI server and get the response.
```
```suggestion
// Client issues requests to a CLI server and gets the response.
```

antoninbas commented on pkg/antctl/client.go:
```diff
@@ -0,0 +1,126 @@
+func (c *Client) Do(cmd *cobra.Command, opt *RequestOption) (io.Reader, error) {
+	kubeconfig, err := c.resolveKubeconfig(opt)
+	if err != nil {
+		klog.Infoln("No kubeconfig, use local mode")
```
```suggestion
		klog.Infoln("No kubeconfig, using local mode")
```

antoninbas commented on pkg/antctl/client.go:
```diff
@@ -0,0 +1,126 @@
+	return &buf, nil
+}
+
+// Do makes the request by the command and the request option.
```
```suggestion
// Do issues the appropriate request to the CLI server using the provided command and request options.
```

antoninbas commented on pkg/antctl/client.go:
```diff
@@ -0,0 +1,126 @@
+// Do makes the request by the command and the request option.
+func (c *Client) Do(cmd *cobra.Command, opt *RequestOption) (io.Reader, error) {
+	kubeconfig, err := c.resolveKubeconfig(opt)
+	if err != nil {
```
nit: Is this the right test here? Is it possible that I have a malformed kubeconfig file, in which case we should probably return an error instead of falling back to local mode.

weiqiangt replied:
Make sense. I changed to you said. See `pkg/antctl/client.go:126` and `pkg/antctl/client.go:94`.

antoninbas commented on pkg/antctl/commandbundle.go:
```diff
@@ -0,0 +1,140 @@
+	"github.com/vmware-tanzu/antrea/pkg/monitor"
+)
+
+// CommandBundle organizes CommandOptions and applies them to a antctl client or server.
```
```suggestion
// CommandBundle organizes CommandOptions and applies them to an antctl client or server.
```

antoninbas commented on pkg/antctl/commandoption.go:
```diff
@@ -0,0 +1,350 @@
+}
+
+// CommandOption defines options to create a cobra.Command for antctl.
+type CommandOption struct {
```
I find the name `CommandOption` confusing. I think it should be one of `Command`, `CommandOptions` or `CommandDefinition`. My favorite is the last one.

weiqiangt replied:
OK, `CommandDefinition `.

antoninbas commented on pkg/antctl/commandbundle.go:
```diff
@@ -0,0 +1,140 @@
+	return path.Join("/apis", b.GroupVersion.Group, b.GroupVersion.Version)
+}
+
+// ApplyToMux applies the CommandBundle to a mux.
```
This command is not useful. I think you want to say that for each command you, add the appropriate handler function to the mux (path `<api prefix>/<cmd>`), along with one discovery handler (path `<api prefix>`)...

weiqiangt replied:
Updated, `pkg/antctl/commandlist.go:55`, PTAL.

antoninbas commented on pkg/antctl/commandbundle.go:
```diff
@@ -0,0 +1,140 @@
+		l := v1.APIResourceList{
+			TypeMeta:     v1.TypeMeta{Kind: "APIResourceList", APIVersion: v1.SchemeGroupVersion.Version},
+			GroupVersion: b.GroupVersion.String(),
+		}
```
probably a stupid question but: what is the point of this if we do populate the `APIResources` list?

weiqiangt replied:
If we do not return a valid APIResourceList, the ServiceDiscovery of the API Server will failed, the `kubectl api-resources` will also not work. But since the APIResource is not for non
ResourceURL which all antctl commands are, I did not populate the the list.

antoninbas commented on pkg/antctl/client.go:
```diff
@@ -0,0 +1,126 @@
+	"k8s.io/klog"
+)
+
+var isAgent = strings.HasPrefix(os.Getenv("POD_NAME"), "antrea-agent")
```
does that mean we can only run `antctl` for the agent from the Pod itself? I thought we would be able to do it from the Node using `/var/run/antrea/antctl.sock`.

weiqiangt replied:
Yes, to cooperate with agent, antctl should be run inside the agent pod. If we run antctl on node, it is hard to tell whether the endpoint of the controller or the node agent the antctl should use.

antoninbas commented on pkg/antctl/commandoption.go:
```diff
@@ -0,0 +1,350 @@
+// CommandOption defines options to create a cobra.Command for antctl.
+type CommandOption struct {
+	// Cobra related
+	Use     string
```
maybe worth pointing out that this is used for the API endpoint

weiqiangt replied:
Done.

antoninbas commented on pkg/antctl/commandoption.go:
```diff
@@ -0,0 +1,350 @@
+		}
+		return err
+	}
+	if co.AddonTransform == nil { // Decode the data if there is no AddonTransform.
```
it's a bit strange that this case is separated from the `co.AddonTransform != nil` above. Is there a reason for that?

weiqiangt replied:
Strange, indeed. I re-organzied the logic, PTAL.

antoninbas commented on pkg/antctl/commandoption.go:
```diff
@@ -0,0 +1,350 @@
+		return fmt.Errorf("error when decoding response: %w", err)
+	}
+	// Output structure data in format
+	format, err := cmd.Flags().GetString("output")
```
if this is the only place you need to use `cmd` here, I don't think you should pass the cobra command as a parameter, just pass the output format directly (and maybe define an enum for it?).

weiqiangt replied:
Refactored. The signature is `resp io.Reader, writer io.Writer, ft formatterType, single bool` now. I also created enums of output formats.

antoninbas commented on pkg/antctl/commandoption.go:
```diff
@@ -0,0 +1,350 @@
+	return result, nil
+}
+
+func (co *CommandOption) format(cmd *cobra.Command, r io.Reader, w io.Writer, single bool) (err error) {
```
worth adding a comment here, for the sake of the developers who will have to work on this in the future. I suppose it is used to parse the reply from the server and output it to the user in the desired format?

weiqiangt replied:
Added, pkg/antctl/commanddefinition.go:295

antoninbas commented on pkg/antctl/definitions.go:
```diff
@@ -0,0 +1,70 @@
+}
+
+// Definition defines command related options of the antctl.
+var Definition = &CommandBundle{
```
should this be just `CommandList` instead of `Definition`?

weiqiangt replied:
Or how about just call it `Bundle`?

weiqiangt replied:
Changed to `commandList`.

weiqiangt commented on test/e2e/infra/vagrant/push_antrea.sh:
```diff
@@ -8,7 +8,9 @@ THIS_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
-ANTREA_YML=$THIS_DIR/../../../../build/yamls/antrea.yml
+ROOT_DIR=$THIS_DIR/../../../..
+ANTREA_YML=$ROOT_DIR/build/yamls/antrea.yml
+ANTCTL=$ROOT_DIR/bin/antctl
```
My fault, duplicated. I'd like to remove codes in `push_antrea.sh`
`antctl` will still be built by the `bin` target, IMO, `antctl-<OS>` are mainly for release or dev.

weiqiangt commented on test/e2e/antctl_test.go:
```diff
@@ -0,0 +1,114 @@
+		{name: "CommandDebug", hasStderr: true, commands: []string{"antctl", "version", "-v"}},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			podName, err := data.getAntreaController()
```
Thanks, moved to the outside of the loop.

weiqiangt commented on test/e2e/antctl_test.go:
```diff
@@ -0,0 +1,114 @@
+	_, _, _, err = RunCommandOnNode(masterNodeName(), cmd)
+	assert.Nil(t, err, "Error when copying antctl from %s", podName)
+
+	_, stdout, stderr, err := RunCommandOnNode(masterNodeName(), "~/antctl -v version")
```
The `antctl` is copied from pod. From the stderr and stdout, I think this test runs as I expected.
But you are right, I will check return code in future.
Just added some return code assertions.

antoninbas replied:
Thanks, I completely missed that the `antclt` binary was built by `make bin`, so I was very confused by what was going on.

That makes the `antctl` target confusing to the user though, because there is a target (`bin`) that produces a binary with this name but it's just not that one. For a given OS `X`, we also end up with 2 identical binaries with different names: `antrea-X` and `antrea`. Documentation may help there, but any thoughts on making this more consistent?

weiqiangt replied:
In my thought, the `antctl-X` is mainly for releases, but I don't have a good idea about where to put them.

weiqiangt replied:
Use #312 to track this issue.

weiqiangt commented:
@tnqn, could you please take a look [this check](https://github.com/vmware-tanzu/antrea/pull/208/checks?check_run_id=361955381). It seems like a similar issue of #111.
cc @mengdie-song.
> \+ /go/bin/mockgen -copyright_file hack/boilerplate/license_header.raw.txt -destination pkg/monitor/testing/mock_monitor.go -package=testing github.com/vmware-tanzu/antrea/pkg/monitor AgentQuerier,ControllerQuerier
92
go: finding github.com/vmware-tanzu/antrea/pkg latest
93
go: finding github.com/vmware-tanzu/antrea/pkg/client latest
94
go: finding github.com/vmware-tanzu/antrea/pkg/client/clientset/versioned latest
95
go: finding github.com/vmware-tanzu/antrea/pkg/client/clientset latest
96
build command-line-arguments: cannot load github.com/vmware-tanzu/antrea/pkg/client/clientset/versioned: no matching versions for query "latest"
97
go: finding github.com/vmware-tanzu/antrea/pkg/client latest
98
go: finding github.com/vmware-tanzu/antrea/pkg latest
99
go: finding github.com/vmware-tanzu/antrea/pkg/client/clientset/versioned latest
100
go: finding github.com/vmware-tanzu/antrea/pkg/client/clientset latest
101
build command-line-arguments: cannot load github.com/vmware-tanzu/antrea/pkg/client/clientset/versioned: no matching versions for query "latest"
102
go: finding github.com/vmware-tanzu/antrea v0.2.0
103
go: downloading github.com/vmware-tanzu/antrea v0.2.0
104
go: extracting github.com/vmware-tanzu/antrea v0.2.0
105
go: github.com/vmware-tanzu/antrea@v0.2.0 requires
106
	github.com/contiv/ofnet@v0.0.0-00010101000000-000000000000: invalid version: unknown revision 000000000000
107
2019/12/24 06:45:53 Loading input failed: exit status 1
108

tnqn commented:
> @tnqn, could you please take a look [this check](https://github.com/vmware-tanzu/antrea/pull/208/checks?check_run_id=361955381). It seems like a similar issue of #111.
> cc @mengdie-song.
> 

Did you change go.mod by intention?

weiqiangt commented:
> > @tnqn, could you please take a look [this check](https://github.com/vmware-tanzu/antrea/pull/208/checks?check_run_id=361955381). It seems like a similar issue of #111.
> > cc @mengdie-song.
> 
> Did you change go.mod by intention?

I only added one dependency `structtag` which looks like not related to this failure. The `gobgp` replace statement is a try to resolve the `codegen` failure.
In this CL, a new code generation `antrea/pkg/monitor` which depends on the `ofnet` is going to be added. The `ofnet` also uses a dummy version number in `require` part in go.mod and has a corresponding `replace` statement.
I'm not sure where the dependency `github.com/vmware-tanzu/antrea v0.2.0` comes from?

mengdie-song commented:
> 
> 
> > > @tnqn, could you please take a look [this check](https://github.com/vmware-tanzu/antrea/pull/208/checks?check_run_id=361955381). It seems like a similar issue of #111.
> > > cc @mengdie-song.
> > 
> > 
> > Did you change go.mod by intention?
> 
> I only added one dependency `structtag` which looks like not related to this failure. The `gobgp` replace statement is a try to resolve the `codegen` failure.
> In this CL, a new code generation `antrea/pkg/monitor` which depends on the `ofnet` is going to be added. The `ofnet` also uses a dummy version number in `require` part in go.mod and has a corresponding `replace` statement.
> I'm not sure where the dependency `github.com/vmware-tanzu/antrea v0.2.0` comes from?
I think that it is because you change go.mod github.com/contiv/ofnet version to 000000000000 in your patch, change it back may solve the issue.

mengdie-song commented:
> 
> 
> > > @tnqn, could you please take a look [this check](https://github.com/vmware-tanzu/antrea/pull/208/checks?check_run_id=361955381). It seems like a similar issue of #111.
> > > cc @mengdie-song.
> > 
> > 
> > Did you change go.mod by intention?
> 
> I only added one dependency `structtag` which looks like not related to this failure. The `gobgp` replace statement is a try to resolve the `codegen` failure.
> In this CL, a new code generation `antrea/pkg/monitor` which depends on the `ofnet` is going to be added. The `ofnet` also uses a dummy version number in `require` part in go.mod and has a corresponding `replace` statement.
> I'm not sure where the dependency `github.com/vmware-tanzu/antrea v0.2.0` comes from?

@weiqiangt Have you tried "make codegen" with your change in the local set up?

vicky-liu commented on cmd/antrea-controller/controller.go:
```diff
@@ -89,10 +95,17 @@ func run(o *Options) error {
+	antctl.CommandList.ApplyToMux(preparedAPIServer.GenericAPIServer.Handler.NonGoRestfulMux, nil, controllerMonitor)
+	go preparedAPIServer.Run(stopCh)
+
+	antctlStoppedCh := antctl.StartServer(antctlServerOpt, antctl.CommandList, nil, controllerMonitor, stopCh)
```
Why we need the input parameter "antctl.CommandList", we cannt get it from antctl directly?

weiqiangt replied:
Make sense, I will remove this parameter.

vicky-liu commented on cmd/antrea-agent/agent.go:
```diff
@@ -133,6 +139,8 @@ func run(o *Options) error {
 
 	go agentMonitor.Run(stopCh)
 
+	antctl.StartServer(antctlServerOpt, antctl.CommandList, agentMonitor, nil, stopCh)
```
Why here we don't introduce a new antctl stop channel to graceful exit?

weiqiangt replied:
It did do like you say, but folks would like to have a consistent solution about the stop channel.

vicky-liu commented on cmd/antctl/main.go:
```diff
@@ -0,0 +1,60 @@
+var rootCmd = &cobra.Command{
+	Use:   commandName,
+	Short: commandName + " is the command line tool for Antrea",
+	Long:  commandName + " is the command line tool for Antrea that supports showing runtime information of ${component}",
```
Maybe changed to a generic message, like status query and configuration setting, as we might introduce read-write cmd.

weiqiangt replied:
Changed to`antctl is the command line tool for Antrea that supports querying status and setting configuration of ${component}`. How about this?

weiqiangt commented:
> > > > @tnqn, could you please take a look [this check](https://github.com/vmware-tanzu/antrea/pull/208/checks?check_run_id=361955381). It seems like a similar issue of #111.
> > > > cc @mengdie-song.
> > > 
> > > 
> > > Did you change go.mod by intention?
> > 
> > 
> > I only added one dependency `structtag` which looks like not related to this failure. The `gobgp` replace statement is a try to resolve the `codegen` failure.
> > In this CL, a new code generation `antrea/pkg/monitor` which depends on the `ofnet` is going to be added. The `ofnet` also uses a dummy version number in `require` part in go.mod and has a corresponding `replace` statement.
> > I'm not sure where the dependency `github.com/vmware-tanzu/antrea v0.2.0` comes from?
> 
> @weiqiangt Have you tried "make codegen" with your change in the local set up?

The `make codegen` works in my local. Maybe it is related to the Github CI, I will try to triage it on my fork.

mengdie-song commented on build/yamls/base/antctl.yml:
```diff
@@ -0,0 +1,32 @@
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: antrea-agent
```
antctl or antrea-antctl ?

weiqiangt replied:
Thanks, antctl.

mengdie-song commented on cmd/antrea-controller/controller.go:
```diff
@@ -89,7 +95,11 @@ func run(o *Options) error {
 
-	go apiServer.GenericAPIServer.PrepareRun().Run(stopCh)
+	preparedAPIServer := apiServer.GenericAPIServer.PrepareRun()
+	antctl.CommandList.ApplyToMux(preparedAPIServer.GenericAPIServer.Handler.NonGoRestfulMux, nil, controllerMonitor)
```
Maybe add a comment here to show that it is for remote mode? Also, you can point out that CLI supports remote mode for controller here.

weiqiangt replied:
Added.

mengdie-song commented on pkg/antctl/commandlist.go:
```diff
@@ -0,0 +1,150 @@
+		klog.Infof("Adding cli handler %s", reqPath)
+		mux.HandleFunc(reqPath, handler)
+	}
+	// Set up the discovery handler.
```
Since it seems to be a key function for antctl, we can enhance the comments for this one. Maybe change the comment to "Set up the dummy discovery handler" in this line to be consistent. "satisfy the apiserver" sounds vague and it will be better to add what the API Server requirements are.

weiqiangt replied:
Updated, PTAL.

mengdie-song commented on pkg/antctl/antctl.go:
```diff
@@ -0,0 +1,140 @@
+			close(s.stoppedCh)
+		}()
+	})
+	return s.stoppedCh
```
Not sure if I miss anything here, but it seems that the return channel is never used? Then why we need to introduce this channel?

weiqiangt replied:
Thanks, removed.

weiqiangt commented:
> The `make codegen` works in my local. Maybe it is related to the Github CI, I will try to triage it on my fork.


I can confirm it is due to the dependencies of generations. The `ci/check-codegen.sh` removes all generations first, but the monitor mocks depend on clientset generations. Since the absence of the clientset during the check, Go will try to checkout code from Github and then it failed due to version matching issue.
It is not straightforward to resolve the dependency cycle at the current time, because the k8s codegen has not supported go module yet. I wrote simple mocks for AgentQuerier and ControllerQuerier.

weiqiangt commented on test/e2e/antctl_test.go:
```diff
@@ -0,0 +1,114 @@
+	tb.Logf("antctl stderr:\n%s", stderr)
+}
+
+func runAntctl(podName string, subCmds []string, data *TestData, tb testing.TB) (string, string, error) {
```
Added.

weiqiangt commented:
/test-all

mengdie-song commented:
> 
> 
> > The `make codegen` works in my local. Maybe it is related to the Github CI, I will try to triage it on my fork.
> 
> I can confirm it is due to the dependencies of generations. The `ci/check-codegen.sh` removes all generations first, but the monitor mocks depend on clientset generations. Since the absence of the clientset during the check, Go will try to checkout code from Github and then it failed due to version matching issue.
> It is not straightforward to resolve the dependency cycle at the current time, because the k8s codegen has not supported go module yet. I wrote simple mocks for AgentQuerier and ControllerQuerier.
> 
> FYI, @mengdie-song @tnqn .



> 
> 
> > The `make codegen` works in my local. Maybe it is related to the Github CI, I will try to triage it on my fork.
> 
> I can confirm it is due to the dependencies of generations. The `ci/check-codegen.sh` removes all generations first, but the monitor mocks depend on clientset generations. Since the absence of the clientset during the check, Go will try to checkout code from Github and then it failed due to version matching issue.
> It is not straightforward to resolve the dependency cycle at the current time, because the k8s codegen has not supported go module yet. I wrote simple mocks for AgentQuerier and ControllerQuerier.
> 
> FYI, @mengdie-song @tnqn .
Thanks for the investigation. So you mean that CI for check code-gen now will broke as long as we introduce a new mock class? I have taken a look at the code of update-codegen-dockerized.sh and it will run mockgen first, then client-gen. So I am thinking that changing orders and put client-gen first will solve the problem.

mengdie-song commented:
> 
> 
> > The `make codegen` works in my local. Maybe it is related to the Github CI, I will try to triage it on my fork.
> 
> I can confirm it is due to the dependencies of generations. The `ci/check-codegen.sh` removes all generations first, but the monitor mocks depend on clientset generations. Since the absence of the clientset during the check, Go will try to checkout code from Github and then it failed due to version matching issue.
> It is not straightforward to resolve the dependency cycle at the current time, because the k8s codegen has not supported go module yet. I wrote simple mocks for AgentQuerier and ControllerQuerier.
> 
> FYI, @mengdie-song @tnqn .
@weiqiangt  Thanks for the investigation. So you mean that CI for check code-gen now will broke as long as we introduce a new mock class? I have taken a look at the code of update-codegen-dockerized.sh and it will run mockgen first, then client-gen. So I am thinking that changing orders and put client-gen first will solve the problem.

weiqiangt commented:
> > @weiqiangt  Thanks for the investigation. So you mean that CI for check code-gen now will broke as long as we introduce a new mock class? I have taken a look at the code of update-codegen-dockerized.sh and it will run mockgen first, then client-gen. So I am thinking that changing orders and put client-gen first will solve the problem.

Great, this works.

jianjuns commented on build/yamls/base/antctl.yml:
```diff
@@ -0,0 +1,32 @@
+---
```
Will antctl.yml be incorporated into antrea.yml or distributed separately? Did not see either with this PR.

weiqiangt replied:
Thanks, I forgot to add it to the kustomize.

jianjuns commented on cmd/antctl/main.go:
```diff
@@ -0,0 +1,60 @@
+var rootCmd = &cobra.Command{
+	Use:   commandName,
+	Short: commandName + " is the command line tool for Antrea",
+	Long:  commandName + " is the command line tool for Antrea that supports querying status and setting configuration of ${component}",
```
As we do not support update operation now, maybe just say "supports showing status of".

weiqiangt replied:
Sure, changed.

jianjuns commented on cmd/antctl/main.go:
```diff
@@ -0,0 +1,60 @@
+var (
+	commandName = path.Base(os.Args[0])
+	inPod       = len(os.Getenv("POD_NAME")) != 0
+	isAgent     = strings.HasPrefix(os.Getenv("POD_NAME"), "antrea-agent")
```
This is fine for now. But later with Windows, we need to support Agent running as a daemon in the host.

jianjuns commented on docs/antctl.md:
```diff
@@ -0,0 +1,66 @@
+# Antctl
```
Could we make a separate PR or at least a separate commit for the documentation?

weiqiangt replied:
OK, I removed this markdown from this CL. It will be submitted separately in another CL.

jianjuns commented on pkg/antctl/antctl.go:
```diff
@@ -0,0 +1,142 @@
+// Copyright 2019 Antrea Authors
```
Should we name the file antctl_server if it is mainly for server code?
Of if there are common code for both server and client, should we move server only code to another file?

weiqiangt replied:
Make sense, moved server related to `server.go`.

jianjuns commented on pkg/antctl/antctl.go:
```diff
@@ -0,0 +1,142 @@
+
+type versionResponse struct {
+	handlers.ComponentVersionResponse `json:",inline" yaml:",inline"`
+	AntctlVersion                     string `json:"antctlVersion" yaml:"antctlVersion"`
```
To make it more explicit, consider name the struct transformedVersionResponse.

weiqiangt replied:
Done.

jianjuns commented on pkg/antctl/commanddefinition.go:
```diff
@@ -0,0 +1,408 @@
+// Copyright 2019 Antrea Authors
```
command_definition.go?

weiqiangt replied:
Done.

jianjuns commented on pkg/antctl/commanddefinition.go:
```diff
@@ -0,0 +1,408 @@
+	// Singleton should be true if the handler always returns single object. The
+	// antctl assumes the response data as a slice of the objects by default.
+	Singleton bool
+	// antctl will not add formatter option for this command if it is true.
```
Do we have such cases? In general I hope antctl to consume normal Controller/Agent API which could be potentially consumed by other consumers, and so formatting to CLI output should be done by anctl.

weiqiangt replied:
I have an ongoing CL for `support-bundle` command which only shows a line of bundle path as result. Although the output of the antctl is not formatted, the server side response is still in JSON format.
On a second thought the plaintext is needed, because formatted results are easier to be parsed. I think I should remove this option.

weiqiangt replied:
Removed.

jianjuns commented on pkg/antctl/commanddefinition.go:
```diff
@@ -0,0 +1,408 @@
+	Agent bool
+	// Singleton should be true if the handler always returns single object. The
+	// antctl assumes the response data as a slice of the objects by default.
+	Singleton bool
```
Not sure singleton is the right term here. Maybe singleObject or isList?

weiqiangt replied:
Changed to `SingleObject`.

jianjuns commented on pkg/antctl/commanddefinition.go:
```diff
@@ -0,0 +1,408 @@
+	// ResponseStruct is the response object data struct of the command. This
+	// struct is no need to be used for the response struct of the handler if
+	// a AddonTransform is set.
+	ResponseStruct interface{}
```
TransformedResponse to be explicit.

weiqiangt replied:
Done.

tnqn commented on pkg/antctl/client.go:
```diff
@@ -0,0 +1,134 @@
+
+// resolveKubeconfig tries to load the kubeconfig specified in the RequestOption.
+// It will return error if the stating of the file failed or the kubeconfig is malformed.
+// It will not try to look up InCluster configuration. If the kubeconfig is loaded,
```
Why don't try InCluster config?

weiqiangt replied:
It was not considered in current implementation, but I'm going to implement an image to run antctl in pod.

weiqiangt replied:
Use #310 to track this issue.

tnqn commented on pkg/antctl/client.go:
```diff
@@ -0,0 +1,134 @@
+// the groupVersion of the RequestOption which will be used to construct the request
+// URI. localRequest is basically a raw http request, none authentication and authorization
+// will be done during the request. For safety concerns, it communicates with the
+// antctl server by a predefined unix domain socket. If the request succeeded, it
```
```suggestion
// antctl server by a predefined unix domain socket. If the request succeeds, it
```

tnqn commented on pkg/antctl/client.go:
```diff
@@ -0,0 +1,134 @@
+// URI. localRequest is basically a raw http request, none authentication and authorization
+// will be done during the request. For safety concerns, it communicates with the
+// antctl server by a predefined unix domain socket. If the request succeeded, it
+// will returns an io.Reader which contains the response data.
```
```suggestion
// will return an io.Reader which contains the response data.
```

tnqn commented on pkg/antctl/client.go:
```diff
@@ -0,0 +1,134 @@
+// data.
+func (c *client) Request(opt *RequestOption) (io.Reader, error) {
+	if c.inPod {
+		klog.Infoln("Antctl runs as local mode")
```
As antctl is the binary name, maybe don't capitalize it?

tnqn commented on pkg/antctl/command_list.go:
```diff
@@ -0,0 +1,170 @@
+// validate checks the validation of the commandList.
+func (cl *commandList) validate() []error {
+	var errs []error
+	if len(cl.definitions) == 0 { // must has at least one applySubCommandToRoot
```
```suggestion
	if len(cl.definitions) == 0 { // must have at least one applySubCommandToRoot
```
But I think the code is self-explaining already, so do L151, no need to have those comments.

tnqn commented on pkg/antctl/handlers/interface.go:
```diff
@@ -0,0 +1,33 @@
+	// running in the antrea-agent, the AgentQuerier will not be nil, otherwise,
+	// the ControllerQuerier will not be nil. If the command has no AddonTransform,
+	// the HandlerFunc need to write the data to the response body in JSON format.
+	Handler(aq monitor.AgentQuerier, cq monitor.ControllerQuerier) http.HandlerFunc
```
AgentQuerier and ControllerQuerier are never provided together, feel not suitable to be the parameters for a generic interface, and can they fulfill all command? Or is it a common way to generate handler? I'm thinking injecting the dependencies via struct fields and leave the interface generic.

weiqiangt replied:
The interface is a factory, I think it is OK to put them here.

tnqn commented:
Besides, I see server side code is still under pkg/antctl, do you plan to change it as @jianjuns suggested, or leave it to later improvement, or we want to keep it as is?

weiqiangt commented on cmd/antctl/main.go:
```diff
@@ -0,0 +1,61 @@
+
+func init() {
+	// prevent any unexpected output at beginning
+	flag.Set("logtostderr", "true")
```
Codes before `ApplyToRootCommand` in `main` may output logs from dependencies. But I found the `alsologtostderr` flag is not needed, removed.

weiqiangt commented:
> Besides, I see server side code is still under pkg/antctl, do you plan to change it as @jianjuns suggested, or leave it to later improvement, or we want to keep it as is?

I would like to refactor this in another PR.

weiqiangt commented:
/test-all

weiqiangt commented:
Hi folks, thanks a lot for reviewing this PR. I have addressed all comments and created issues to track several future works of it. The document for the antctl will also be added in a separate PR(it was picked out from this PR).

antoninbas commented on pkg/antctl/command_definition.go:
```diff
@@ -0,0 +1,425 @@
+	// GroupVersion is the group version of the command handler, it should be set
+	// alongside with the HandlerFactory.
+	GroupVersion *schema.GroupVersion
+	// TransformedResponse is the final response struct of the command. If the
```
I am not sure I understand this comment fully, but I think what you mean is:
> TransformedResponse is used to identify the struct type of the response received from the server (and that we need to output in the desired format). This type information is used to convert the response received from the server to the appropriate struct. If AddonTransform is set, TransformedResponse is not needed as AddonTransform is in-charge of converting the response to a struct.

If the description above is correct, have you considered using a `reflect.Type` here instead if an `interface{}`?

weiqiangt replied:
Sorry I didn't clarify it. The `TransformedResponse` is always be used to tell the formatter how to output the response. The `AddonTransform` is used to parse the response from a handler but not is not for outputting.

The suggestion is good, I have updated the code accordingly.

antoninbas replied:
Thanks for clarifying. It seems that the original comment implied incorrectly that the `TransformedResponse` was sometimes not needed.

antoninbas commented on pkg/antctl/command_definition.go:
```diff
@@ -0,0 +1,425 @@
+	return ret
+}
+
+// decode parses the data in reader to the TransformedResponse. If the single is false,
```
I suggest the following:
> decode parses the data in reader and converts it to one or more TransformedResponse objects. If single is false, the return type is []TransformedResponse. Otherwise, the return type is TransformedResponse.

weiqiangt replied:
Updated.

antoninbas commented on pkg/antctl/antctl.go:
```diff
@@ -0,0 +1,80 @@
+			Long:                "Print version information of the antctl and the ${component}",
+			HandlerFactory:      new(handlers.Version),
+			GroupVersion:        &systemGroup,
+			TransformedResponse: new(transformedVersionResponse),
```
If I understand the code correctly, this could be nil because we have an AddonTransform function?

weiqiangt replied:
It should not be nil, this field is used to tell the output structure. If it is nil, the framework would not know how to format the result.
The workflow would look like:
```
(handler response struct) -> [default transform] -> (transformed response struct) -> [format]
                          \> [add-on transform] / 
```

antoninbas replied:
Understood, I was confused because of https://github.com/vmware-tanzu/antrea/pull/208/files#r366599836

antoninbas commented on pkg/antctl/handlers/version.go:
```diff
@@ -0,0 +1,58 @@
+// command. It contains the version of the component the antctl server is running
+// in, either the agent or the controller.
+// This struct is not the final response struct of the version command. The version
+// command definition has a AddonTransform which will populate this struct and the
```
```suggestion
// command definition has an AddonTransform which will populate this struct and the
```

antoninbas commented on test/e2e/antctl_test.go:
```diff
@@ -0,0 +1,144 @@
+	return stdout, stderr, err
+}
+
+// TestAntctlAgentLocalAccess ensures antctl is accessible in agent pod.
```
s/pod/Pod (here and other places in the file)

weiqiangt replied:
Done.

antoninbas commented on test/e2e/antctl_test.go:
```diff
@@ -0,0 +1,144 @@
+	podName, err := data.getAntreaController()
+	assert.Nil(t, err, "Error when retrieving antrea controller pod name")
+
+	// Copy antctl from the controller pod to the master node.
```
s/node/Node (here and other places)

weiqiangt replied:
Done.

antoninbas commented on test/e2e/antctl_test.go:
```diff
@@ -0,0 +1,144 @@
+	}
+	defer teardownTest(t, data)
+	podName, err := data.getAntreaController()
+	assert.Nil(t, err, "Error when retrieving antrea controller pod name")
```
here and in other places below, `require` may be more appropriate than `assert`. If we cannot retrieve the Pod name here, we probably want to interrupt the test.

weiqiangt replied:
Thanks, I expected `assert` just works like it does in other languages. Updated.

weiqiangt commented on pkg/antctl/command_definition.go:
```diff
@@ -0,0 +1,425 @@
+
+// applySubCommandToRoot applies the commandDefinition to a cobra.Command with
+// the client. It populates basic fields of a cobra.Command and creates specific
+// the RunE function for it according the commandDefinition.
```
Updated.

weiqiangt commented on pkg/antctl/command_definition.go:
```diff
@@ -0,0 +1,425 @@
+	return result, nil
+}
+
+func (cd *commandDefinition) plainTextOutput(resp io.Reader, writer io.Writer) error {
```
Sorry, it is legacy codes, removed.

weiqiangt commented on pkg/antctl/command_definition.go:
```diff
@@ -0,0 +1,425 @@
+
+// output reads bytes from the resp and outputs the data to the writer in desired
+// format. If the AddonTransform is set, it will use the function to transform
+// the data first. If the commandDefinition marked the response bytes as PlainText,
```
removed.

weiqiangt commented:
/test-all

jianjuns commented:
Again, I believe we should decouple API server from antctl, but we can merge the current implementation, and change later (maybe after commands for runtime info are done).

weiqiangt commented:
> Again, I believe we should decouple API server from antctl, but we can merge the current implementation, and change later (maybe after commands for runtime info are done).

Copied.
